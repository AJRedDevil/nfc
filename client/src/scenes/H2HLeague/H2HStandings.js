// npm packages
import React from 'react';

// our packages
import Animation from '../../components/animations';
import Table from '../../components/Table';
import H2HStandingsPropTypes from './PropTypes';

const getBody = standings =>
  standings.map(item => [
    item.rank,
    item.entry_name,
    item.player_name,
    item.matches_won,
    item.matches_drawn,
    item.matches_lost,
    item.points_for,
    item.points_total,
  ]);

const H2HStandings = props => (
  <Animation.Fade in>
    <div>
      <Table body={getBody(props.standings)} head={props.head} />
    </div>
  </Animation.Fade>
);
H2HStandings.propTypes = {
  head: H2HStandingsPropTypes.head,
  standings: H2HStandingsPropTypes.body,
};

export default H2HStandings;
