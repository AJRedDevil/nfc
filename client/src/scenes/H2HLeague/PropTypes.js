import PropTypes from 'prop-types';

const head = PropTypes.arrayOf(PropTypes.shape({text: PropTypes.string}));

const body = PropTypes.arrayOf(PropTypes.object);

const h2hLeagueSchema = PropTypes.shape(head);

const data = PropTypes.shape({
  lastFetched: PropTypes.string,
  leagueNames: PropTypes.arrayOf(PropTypes.string),
  standings: PropTypes.arrayOf(PropTypes.arrayOf(PropTypes.object)),
});

export default {h2hLeagueSchema, data, head, body};
