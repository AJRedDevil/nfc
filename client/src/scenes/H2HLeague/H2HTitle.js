// npm packages
import React, {Component} from 'react';
import PropTypes from 'prop-types';

// our packages
import LinkTab from '../../components/LinkTab';

class H2HTitle extends Component {
  state = {currentTab: this.props.links[0].path};

  handleTabClick = e => {
    const currentTab = e.target.getAttribute('identifier');
    this.setState({
      ...this.state,
      currentTab,
    });
  };

  render() {
    const {links} = this.props;
    return (
      <div className="tabs">
        <ul>
          {links &&
            links.map(link => (
              <LinkTab
                key={link.text}
                link={link}
                currentTab={this.state.currentTab}
                onTabClick={this.handleTabClick}
              />
            ))}
        </ul>
      </div>
    );
  }
}
H2HTitle.propTypes = {
  links: PropTypes.arrayOf(
    PropTypes.shape({
      path: PropTypes.string,
      text: PropTypes.string,
    })
  ),
};

export default H2HTitle;
