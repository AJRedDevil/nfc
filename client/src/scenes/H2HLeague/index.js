// npm packages
import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {isEmpty} from 'lodash';
import {connect} from 'react-redux';
import {Route, Switch} from 'react-router-dom';

// our packages
import H2HTitle from './H2HTitle';
import H2HStandings from './H2HStandings';
import H2HLeaguePropTypes from './PropTypes';
import LoadingTable from '../../components/LoadingTable';
import {DateHelper} from '../../utils';
import {fetchH2HStandingsData, setRoute} from '../../services/h2h/actions';

class H2HLeague extends Component {
  componentWillMount() {
    if (this.props.currentTab) {
      const defaultUrl = `${this.props.match.url}/${this.props.currentTab}`;
      this.props.history.push(defaultUrl);
    }
  }

  componentDidMount() {
    const {lastFetched, standings} = this.props.data;
    if (isEmpty(standings) || DateHelper.isAnHourAgo(lastFetched)) {
      this.props.fetchH2HData();
    }
  }

  componentWillReceiveProps(nextProps) {
    if (
      isEmpty(this.props.currentTab) &&
      this.props.currentTab !== nextProps.currentTab
    ) {
      const defaultUrl = `${this.props.match.url}/${nextProps.currentTab}`;
      this.props.history.push(defaultUrl);
    }
  }

  renderStandings = () => {
    const {links, match, data, schema, currentTab} = this.props;
    const routes = links.map(link => (
      <Route
        key={link}
        path={`${match.url}/${link.path}`}
        render={() => (
          <H2HStandings standings={data.standings[currentTab]} {...schema} />
        )}
      />
    ));
    return routes;
  };

  renderH2HStandings = () => (
    <div>
      <H2HTitle {...this.props} />
      <Switch>{this.renderStandings()}</Switch>
    </div>
  );

  render() {
    return isEmpty(this.props.currentTab) ? (
      <LoadingTable title="H2HStandings" />
    ) : (
      this.renderH2HStandings()
    );
  }
}
H2HLeague.propTypes = {
  currentTab: PropTypes.string,
  links: PropTypes.arrayOf(
    PropTypes.shape({
      path: PropTypes.string,
      text: PropTypes.string,
    })
  ),
  schema: H2HLeaguePropTypes.h2hLeagueSchema,
  data: H2HLeaguePropTypes.data,
  fetchH2HData: PropTypes.func.isRequired,
  match: PropTypes.shape({
    isExact: PropTypes.bool,
    params: PropTypes.object,
    path: PropTypes.string,
    url: PropTypes.string,
  }),
  history: PropTypes.object,
  setNextTab: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  currentTab: state.H2HStandings.currentTab,
  links: state.H2HStandings.links,
  data: state.H2HStandings.data,
  schema: state.H2HStandings.h2hLeagueSchema,
});

const mapDispatchToProps = dispatch => ({
  fetchH2HData: () => dispatch(fetchH2HStandingsData()),
  setNextTab: nextTab => dispatch(setRoute(nextTab)),
});

export default connect(mapStateToProps, mapDispatchToProps)(H2HLeague);
