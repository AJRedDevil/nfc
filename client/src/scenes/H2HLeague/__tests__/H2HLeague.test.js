// npm packages
import React from 'react';
import renderer from 'react-test-renderer';
import configureMockStore from 'redux-mock-store';
import {MemoryRouter} from 'react-router-dom';

// our packages
import H2HLeague from '../index';

// create mock store
const mockStore = configureMockStore();

test('# H2HLeague', () => {
  const standings = {
    1: [
      {
        rank: 1,
        entry_name: 'test',
        player_name: 'test_1',
        matches_won: 1,
        matches_drawn: 1,
        matches_lost: 1,
        points_for: 1,
        points_total: 100,
      },
    ],
    2: [
      {
        rank: 2,
        entry_name: 'test',
        player_name: 'test_2',
        matches_won: 2,
        matches_drawn: 2,
        matches_lost: 2,
        points_for: 2,
        points_total: 200,
      },
    ],
  };
  const store = mockStore({
    H2HStandings: {
      currentTab: '1',
      links: [
        {
          path: '',
          text: 'HomePage',
        },
        {
          path: 'test',
          text: 'TestPage',
        },
      ],
      data: {
        lastFetched: '2018-02-20',
        allIds: ['1', '2'],
        leagueNames: {
          1: 'league_1',
          2: 'league_2',
        },
        standings,
      },
    },
    schema: {
      head: [
        {text: 'Rank'},
        {text: 'Team'},
        {text: 'Player'},
        {text: 'Won'},
        {text: 'Drawn'},
        {text: 'Lost'},
        {text: 'For'},
        {text: 'Total'},
      ],
    },
  });
  const match = {
    isExact: false,
    params: {},
    path: 'http://localhost:8080',
    url: '',
  };
  const history = {
    push: () => {},
  };
  const links = [{path: '', text: 'home'}, {path: 'test', text: 'TestPage'}];

  const component = renderer
    .create(
      <MemoryRouter>
        <H2HLeague
          store={store}
          match={match}
          history={history}
          links={links}
        />
      </MemoryRouter>
    )
    .toJSON();
  expect(component).toMatchSnapshot();
});
