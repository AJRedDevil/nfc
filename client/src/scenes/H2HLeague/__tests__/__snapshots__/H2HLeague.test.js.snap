// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`# H2HLeague 1`] = `
<div>
  <div
    className="tabs"
  >
    <ul>
      <li
        className=""
      >
        <a
          aria-current="true"
          className=""
          href="/"
          identifier=""
          onClick={[Function]}
          style={Object {}}
        >
          HomePage
        </a>
      </li>
      <li
        className=""
      >
        <a
          aria-current={false}
          className={undefined}
          href="/test"
          identifier="test"
          onClick={[Function]}
          style={undefined}
        >
          TestPage
        </a>
      </li>
    </ul>
  </div>
  <div>
    <table
      className="table is-fullwidth is-striped is-narrow is-hoverable"
    >
      <thead>
        <tr />
      </thead>
      <tfoot>
        <tr />
      </tfoot>
      <tbody>
        <tr>
          <th>
            1
          </th>
          <td>
            test
          </td>
          <td>
            test_1
          </td>
          <td>
            1
          </td>
          <td>
            1
          </td>
          <td>
            1
          </td>
          <td>
            1
          </td>
          <td>
            100
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
